kotlin/src/main/com/looker/rtl/Transport.kt

```diff
@@ -0,0 +1,202 @@
+        val headers: Map<String, String> = mapOf()
+)
+
+typealias Authenticator = (init: RequestSettings) -> RequestSettings
+
+fun defaultAuthenticator(requestSettings: RequestSettings): RequestSettings = requestSettings
+
+open class TransportSettings(
+        var baseUrl: String = "",
+        var apiVersion: String = "",
+        var headers: Map<String, String> = mapOf())
+
+fun encodeValues(params: Values = mapOf()): String {
+    @Suppress("UNCHECKED_CAST")
+    return params
+            .filter { (k, v) -> v !== null }
+            .map { (k, v) -> "$k=${URLEncoder.encode("$v", "utf-8")}"}
+            .joinToString("&")
+
+}
+
+fun addQueryParams(path: String, params: Values = mapOf()): String {
+    if (params.isEmpty()) return path
+
+    val qp = encodeValues(params)
+    return "$path?$qp"
+}
+
+class Transport(val options: TransportSettings) {
+
+    var httpClient: HttpClient? = null
+        private set
+
+    // Internal only secondary constructor to support supplying an HTTP client for testing
+    internal constructor(options: TransportSettings, httpClient: HttpClient) : this(options) {
+        this.httpClient = httpClient
+    }
+
+    init {
+        if (httpClient == null)
+            httpClient = HttpClient(Apache) {
+                install(JsonFeature) {
+                    serializer = JacksonSerializer()
+                }
+            }
+    }
+
+    val apiPath = "${options.baseUrl}/api/${options.apiVersion}"
+
+    private fun ok(res: HttpResponse): Boolean {
+        // TODO: Should this use an enum class like in the ts version?
+        // Enums in Kotlin don't work like c or ts and would require
+        // an explicit numeric value for each enum type
+        // Thought: We should use whatever is idiomatic for Kotlin
+        return (res.status.value >= 200) && (res.status.value <= 226)
+    }
+
+    /**
+     * Create the correct http request path
+     * @param path Relative or absolute path
+     * @param queryParams query string arguments (if any)
+     * @param authenticator optional authenticator callback for API requests
+     * @return a fully qualified path that is the base url, the api path, or a pass through request url
+     */
+    fun makePath(
+            path: String,
+            queryParams: Values = mapOf(),
+            authenticator: Authenticator? = null
+    ): String {
+        return if (path.startsWith("http://", true)
+                || path.startsWith("https://", true)) {
+            "" // full path was passed in
+        } else {
+            if (authenticator === null)  {
+                options.baseUrl
+            } else {
+                apiPath
+            }
+        } + addQueryParams(path, queryParams)
+    }
+
+    inline fun <reified T> request(
+            method: HttpMethod,
+            path: String,
+            queryParams: Values = mapOf(),
+            body: Any? = null,
+            noinline authenticator: Authenticator? = null): SDKResponse {
```

Comment:
Why did we change the default here from `Authenticator = ::defaultAuthenticator` to make it nullable? Is this due to the needing an `Optional` class to wrap nullable fields?

Reply:
I've added a TODO